boa3_test/tests/boatestcase.py

```diff
@@ -415,6 +421,38 @@ def _unwrap_stack_item(cls, stack_item: noderpc.StackItem, expected_type: type |
             result = stack_item.value
         return result
 
+    def filter_events(self,
+                      events: list[noderpc.Notification],
+                      *,
+                      origin: types.UInt160 | list[types.UInt160] = None,
+                      event_name: str = None,
+                      notification_type: Type[T] = noderpc.Notification
+                      ) -> list[T]:
+
+        if issubclass(notification_type, BoaTestEvent):
+            convert_event: Callable[[noderpc.Notification], BoaTestEvent] = notification_type.from_untyped_notification
+        elif hasattr(notification_type, 'from_notification'):
+            convert_event: Callable[[noderpc.Notification], TestNotification] = notification_type.from_notification
+        else:
+            convert_event: Callable[[noderpc.Notification], noderpc.Notification] = lambda event: event
+
+        if origin is None and event_name is None:
+            return [convert_event(notification) for notification in events]
+
+        if origin is None:
+            return [convert_event(notification) for notification in events if notification.event_name == event_name]
+        elif not isinstance(origin, list):
+            origin = [origin]
+
+        if event_name is None:
+            return [convert_event(notification) for notification in events
+                    if notification.contract in origin
+                    ]
+        else:
+            return [convert_event(notification) for notification in events
+                    if notification.event_name == event_name and notification.contract in origin
+                    ]
```

Comment:
I don't like your suggestion because of how many iterations we need to do.

- origin and event_name None:
  1 copy and 1 list iteration on return
- origin None or event_name None:
  1 copy and 2 list iterations in total
- origin and event_name not None:
  1 copy and 3 list iterations in total

I wrote the method this way thinking in how much time it would take to handle many notification for more complex testing than to being the easier to rewrite if we need.

Reply:
I think it would be possible to attribute functions instead of iterating the list, something like this:
```python
check_origin = lambda notification: True
check_event_name = lambda notification: True

if origin is not None:
    if not isinstance(origin, list):
        origin = [origin]

    check_origin = lambda notification: notification.contract in origin

if event_name:
    check_event_name = lambda notification: notification.event_name == event_name

return [convert_event(notification) for notification in events if check_origin(notification) and check_event_name(notification)]
```

Tho, if we don't need to worry about rewriting this function, let's keep it as it is